lib/ueberauth/strategy.ex

```diff
@@ -318,4 +339,80 @@ defmodule Ueberauth.Strategy do
     auth = apply(strategy, :auth, [conn])
     Plug.Conn.assign(conn, :ueberauth_auth, auth)
   end
+
+  defp state_param_matches?(conn) do
+    state_cookie = get_state_cookie(conn)
+    param_cookie = conn.params["state"]
+
+    case {state_cookie, param_cookie} do
+      {nil, nil} -> false
+      {value, value} -> true
+      {_, _} -> false
+    end
+  end
+
+  defp add_state_mismatch_error(conn, strategy) do
+    conn
+    |> Helpers.set_errors!([
+      %Error{message_key: :csrf_attack, message: "Cross-Site Request Forgery attack"}
+    ])
+    |> run_handle_cleanup(strategy)
+  end
+
+  defp run_handle_request(conn, strategy) do
+    apply(strategy, :handle_request!, [conn])
+  end
+
+  defp run_handle_callback(conn, strategy) do
+    strategy
+    |> apply(:handle_callback!, [conn])
+    |> handle_callback_result(strategy)
+    |> run_handle_cleanup(strategy)
+  end
+
+  defp run_handle_cleanup(conn, strategy) do
+    # TODO: should we remove the cookie always?
+    conn = remove_state_cookie(conn)
+    apply(strategy, :handle_cleanup!, [conn])
+  end
+
+  defp maybe_add_state_param(conn, strategy) do
+    if get_ignores_csrf_attack_option(strategy) do
+      conn
+    else
+      add_state_param(conn)
+    end
+  end
+
+  defp get_ignores_csrf_attack_option(strategy) do
+    strategy
+    |> apply(:default_options, [])
+    |> Keyword.get(:ignores_csrf_attack, false)
+  end
+
+  defp add_state_param(conn) do
+    state = create_state_param()
+
+    conn
+    |> Conn.put_resp_cookie(@state_param_cookie_name, state,
+      # TODO: what would be a good default
+      same_site: "Strict"
+    )
+    |> Helpers.add_state_param(state)
```

Comment:
Why do we add the state here aswell, is the session not closed right after?

Reply:
We use the `Plug.Conn` object from the clients, so we need to read the state param out again, and add the state param to the queries (most likely) of the request.

So this is just a temporary hold so clients can read it out.

Reply:
That makes a lot of sense. Thank you